File: src/prod/src/ServiceModel/hosting2/ContainerImageDescription.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace std;
using namespace Common;
using namespace ServiceModel;
using namespace Hosting2;

ContainerImageDescription::ContainerImageDescription() 
    : imageName_(),
    useDefaultRepositoryCredentials_(),
    useTokenAuthenticationCredentials_(),
    repositoryCredentials_()
{
}

ContainerImageDescription::ContainerImageDescription(
    wstring const & imageName,
    bool useDefaultRepositoryCredentials,
    bool useTokenAuthenticationCredentials,
    RepositoryCredentialsDescription const & repositoryCredentials) 
    : imageName_(imageName),
    useDefaultRepositoryCredentials_(useDefaultRepositoryCredentials),
    useTokenAuthenticationCredentials_(useTokenAuthenticationCredentials),
    repositoryCredentials_(repositoryCredentials)
{
}

void ContainerImageDescription::WriteTo(__in TextWriter & w, FormatOptions const &) const
{
    w.Write("ContainerImageDescription { ");
    w.Write("imageName = {0}, ", imageName_);
    w.Write("useDefaultRepositoryCredentials_ = {0}, ", useDefaultRepositoryCredentials_);
    w.Write("useTokenAuthenticationCredentials_ = {0}, ", useTokenAuthenticationCredentials_);
    w.Write("}");
}

ErrorCode ContainerImageDescription::ToPublicApi(
    __in ScopedHeap & heap,
    __out FABRIC_CONTAINER_IMAGE_DESCRIPTION & fabricContainerImageDesc) const
{
    fabricContainerImageDesc.ImageName = heap.AddString(this->ImageName);

    auto repoCredential = heap.AddItem<FABRIC_REPOSITORY_CREDENTIAL_DESCRIPTION>();
    auto error = this->RepositoryCredentials.ToPublicApi(heap, *repoCredential);
    if (!error.IsSuccess())
    {
        return error;
    }

    fabricContainerImageDesc.RepositoryCredential = repoCredential.GetRawPointer();

    auto fabricContainerImageDescEx1 = heap.AddItem<FABRIC_CONTAINER_IMAGE_DESCRIPTION_EX1>();
    fabricContainerImageDescEx1->UseDefaultRepositoryCredentials = this->UseDefaultRepositoryCredentials;
    fabricContainerImageDesc.Reserved = fabricContainerImageDescEx1.GetRawPointer();

    auto fabricContainerImageDescEx2 = heap.AddItem<FABRIC_CONTAINER_IMAGE_DESCRIPTION_EX2>();
    fabricContainerImageDescEx2->UseTokenAuthenticationCredentials = this->UseTokenAuthenticationCredentials;
    fabricContainerImageDescEx2->Reserved = nullptr;

    fabricContainerImageDescEx1->Reserved = fabricContainerImageDescEx2.GetRawPointer();

    return ErrorCode(ErrorCodeValue::Success);
}

void ContainerImageDescription::SetContainerRepositoryPassword(std::wstring const& password)
{
    repositoryCredentials_.Password = password;
}

File: src/prod/src/managed/Api/src/System/Fabric/Hosting/Container/ContainerImageDescription.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric
{
    using System.Collections.Generic;
    using System.Fabric.Common;
    using System.Fabric.Interop;

    internal sealed class ContainerImageDescription
    {
        internal ContainerImageDescription()
        {
        }

        internal string ImageName { get; set; }

        internal RepositoryCredentialDescription RepositoryCredential { get; set; }

        internal bool UseDefaultRepositoryCredentials { get; set; }

        internal bool UseTokenAuthenticationCredentials { get; set; }

        internal static unsafe ContainerImageDescription CreateFromNative(IntPtr nativePtr)
        {
            ReleaseAssert.AssertIfNot(nativePtr != IntPtr.Zero, "ContainerImageDescription.CreateFromNative() has null pointer.");

            var nativeDescription = *((NativeTypes.FABRIC_CONTAINER_IMAGE_DESCRIPTION*)nativePtr);

            return CreateFromNative(nativeDescription);
        }

        internal static unsafe ContainerImageDescription CreateFromNative(
            NativeTypes.FABRIC_CONTAINER_IMAGE_DESCRIPTION nativeDescription)
        {
            var containerImageDescription =  new ContainerImageDescription
            {
                ImageName = NativeTypes.FromNativeString(nativeDescription.ImageName),
                RepositoryCredential = RepositoryCredentialDescription.CreateFromNative(nativeDescription.RepositoryCredential),
            };

            if (nativeDescription.Reserved != null)
            {
                var nativeParametersEx1 = *((NativeTypes.FABRIC_CONTAINER_IMAGE_DESCRIPTION_EX1*)nativeDescription.Reserved);
                containerImageDescription.UseDefaultRepositoryCredentials = NativeTypes.FromBOOLEAN(nativeParametersEx1.UseDefaultRepositoryCredentials);

                if(nativeParametersEx1.Reserved != null)
                {
                    var nativeParametersEx2 = *((NativeTypes.FABRIC_CONTAINER_IMAGE_DESCRIPTION_EX2*)nativeParametersEx1.Reserved);
                    containerImageDescription.UseTokenAuthenticationCredentials = NativeTypes.FromBOOLEAN(nativeParametersEx2.UseTokenAuthenticationCredentials);
                }
            }

            return containerImageDescription;
        }

        internal static unsafe List<ContainerImageDescription> CreateFromNativeList(IntPtr nativePtr)
        {
            ReleaseAssert.AssertIfNot(nativePtr != IntPtr.Zero, "ContainerImageDescription.CreateFromNativeList() has null pointer.");

            var imageList = new List<ContainerImageDescription>();

            var nativeList = (NativeTypes.FABRIC_CONTAINER_IMAGE_DESCRIPTION_LIST*)nativePtr;

            if(nativeList->Count > 0)
            {
                var items = (NativeTypes.FABRIC_CONTAINER_IMAGE_DESCRIPTION*)(nativeList->Items);

                for (var idx = 0; idx < nativeList->Count; ++idx)
                {
                    imageList.Add(CreateFromNative(items[idx]));
                }
            }

            return imageList;
        }
    }
}